x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,1301 @@
+-   [Exporting and importing
+    rules](#exporting-and-importing-rules){#toc-exporting-and-importing-rules}
+    -   [Exporting rules](#exporting-rules){#toc-exporting-rules}
+    -   [Importing rules](#importing-rules){#toc-importing-rules}
+-   [Customizing Prebuilt Rules](#customizing-prebuilt-rules){#toc-customizing-prebuilt-rules}
+    -   [Endpoints](#endpoints){#toc-endpoints}
+    -   [In the UI](#in-the-ui){#toc-in-the-ui}
+-   [List of things to fix (create tickets)](#list-of-things-to-fix-create-tickets){#toc-list-of-things-to-fix-create-tickets}
+    -   [Rule fields](#rule-fields){#toc-rule-fields}
+-   [Updating Prebuilt Rules](#updating-prebuilt-rules){#toc-updating-prebuilt-rules}
+    -   [Changes to upgrade `_review` and `_perform` endpoints](#changes-to-upgrade-_review-and-_perform-endpoints){#toc-changes-to-upgrade-_review-and-_perform-endpoints}
+    -   [Rule field diff algorithms](#rule-field-diff-algorithms){#toc-rule-field-diff-algorithms}
+-   [Changes in UI](#changes-in-ui){#toc-changes-in-ui}
+-   [Scenarios for bulk accepting updates](#scenarios-for-bulk-accepting-updates){#toc-scenarios-for-bulk-accepting-updates}
+-   [Marking rules (and their fields) as customized](#marking-rules-and-their-fields-as-customized){#toc-marking-rules-and-their-fields-as-customized}
+-   [Other open questions](#other-open-questions){#toc-other-open-questions}
+
+
+
+## Necessary rule schema changes
+
+In order to support the customization of Elastic Prebuilt Rules, we need to modify our rule schema. This involves introducing two new fields: `prebuilt` and `customized`. Simultaneously, we need to deprecate the existing `immutable` field, which determines if a rule is an Elastic Prebuilt Rule and whether its fields can be modified/customized.
+
+To ensure backward compatibility and avoid breaking changes, we will silently deprecate the `immutable` field but keep it unchanged (i.e. non-optional).
+
+
+In summary, the changes to be done to our rule schema are:
+
+- Mark the `immutable` field as deprecated but keep unchanged for backwards compatibility.
+    - Should remain a `boolean` type and non-optional (`required`)
+- Add `prebuilt` field to the rule schema
+    - `prebuilt` should have same type as `immutable`: `boolean`
+    - Should be non-optional (`required`)
+    - During migration, set the `prebuilt` value to match the current value of `immutable`. This means: prebuilt rules should be migrated with a `prebuilt` value set to `true`, while custom rules should be set with a `prebuilt` value of `false`.
+- Add `customized` field to the rule schema
+    - Should have type of `boolean`
+    - Should be non-optional (`required`)
+    - If non existing, initialize it with a default value of `false` for all rules, both custom and prebuilt. The only exception for this is when importing prebuilt rules (see below).
+    - Custom rules (`prebuilt: false`) will never use this field, and its value should always be `false`.
```

Comment:
Did you consider nesting prebuilt-specific fields inside an object? When there's a relationship like this between 2 sibling fields, it's often a good candidate for making them subfields of an object. e.g. something like
```
{
  prebuilt?: {
    customized: boolean;
  }
}
```
This makes it easy later on to add additional fields that are specific to prebuilt rules. We used this pattern recently with suppression and `investigation_fields`.

Reply:
Hi @marshallmain we've given some thought on this with @banderror and liked your proposal, with some modifications. What do you think of the following?

```
{  
  prebuilt: {  
    isPrebuilt: boolean;
    isCustomized: boolean:
    elasticLastUpdated: Date
  }  
}  
```
First of all, we think it makes sense to have a `prebuilt` top-level object where we can group everything related to prebuilt rules. Apart from the fields I wrote above (more on that below), we though we might want to add other related fields in the future, for example a `customizations` field, which could somehow store all user's modifications to a rule's fields (for example, in case we want the user to be able to search rules by their customized field - "In which of my prebuilt rules have I modifed my index patterns field?").

But back to the current design:
- the prebuilt field shouldn't be optional, since we explicitly want to mark if a rule is prebuilt or not. So
- `isPrebuilt`: would be the first nonOptional boolean key, with the same behaviour as the `prebuilt` field as I described up to now in the RFC
- `isCustomized`: would be the second nonOptional boolean key, with the same behaviour as the `customized` field as I described up to now in the RFC
- `elasticLastUpdated`: is an optional field that

Reply:
I think the additional fields (`customizations`, `elasticLastUpdated`, `isCustomized`) make sense, but IMO it would be simpler to leave `isPrebuilt` out entirely, make `prebuilt` optional, and use the existence of `prebuilt` to determine if a rule is prebuilt or not.
```
{  
  prebuilt?: {
    isCustomized: boolean;
    elasticLastUpdated: Date;
  }  
}
```
This makes the schema easier to represent:
```
const prebuilt = z.object({
  isCustomized: z.boolean(),
  elasticLastUpdated: z.date(),
});
type Prebuilt = z.infer<typeof prebuilt>;

const rule = z.object({
  prebuilt: prebuilt.optional(),
  field1: z.string(),
  field2: z.string(),
});
type Rule = z.infer<typeof rule>;
```

If `prebuilt` is required and some subfields take special values based on the value of `isPrebuilt`, then it's a bit more complicated to encode that `isCustomized` and other fields are only valid if `isPrebuilt` is `true`:
```
const prebuiltTrue = z.object({
  isPrebuilt: z.literal(true),
  isCustomized: z.boolean(),
  elasticLastUpdated: z.date(),
});

const prebuiltFalse = z.object({
  isPrebuilt: z.literal(false),
  isCustomized: z.literal(false), // Included per the RFC, but I'd leave this out entirely even if `isPrebuilt` does end up being necessary
});

const prebuilt = z.union([prebuiltTrue, prebuiltFalse]);
type Prebuilt = z.infer<typeof prebuilt>;

const rule = z.object({
  prebuilt,
  field1: z.string(),
  field2: z.string(),
});
type Rule = z.infer<typeof rule>;
```

To me it feels weird to include a `prebuilt` section for rules that are not prebuilt, and it seems sufficient that if `prebuilt` is missing then the rule is implicitly not a prebuilt rule. We take a similar approach with suppression where if the suppression fields are not included then suppression is not enabled for the rule. Do we need to explicitly mark non-prebuilt rules?

Reply:
The data model has been redesigned so I'm resolving this thread.